A non-empty `root_module_direct_deps` disallows users to use the module extension with `dev_dependency = True`

When I attempted to update `hermetic_cc_toolchain` from `3.1.1` to `3.2.0` [here](https://github.com/yuyawk/rules_build_error/pull/106), I saw an error caused by the module extension.

```
2025/02/19 07:23:53 Downloading https://releases.bazel.build/7.5.0/release/bazel-7.5.0-linux-x86_64...
Extracting Bazel installation...
Starting local Bazel server and connecting to it...
Computing main repo mapping: 
Computing main repo mapping: 
Loading: 
Loading: 0 packages loaded
Analyzing: target @@buildifier_prebuilt~//:buildifier (1 packages loaded, 0 targets configured)
Analyzing: target @@buildifier_prebuilt~//:buildifier (1 packages loaded, 0 targets configured)

ERROR: root_module_direct_deps must be empty if the root module contains no usages with dev_dependency = False
ERROR: Analysis of target '@@buildifier_prebuilt~//:buildifier' failed; build aborted: error evaluating module extension toolchains in @@hermetic_cc_toolchain~//toolchain:ext.bzl
INFO: Elapsed time: 5.674s, Critical Path: 0.04s
INFO: 1 process: 1 internal.
ERROR: Build did NOT complete successfully
FAILED: 
ERROR: Build failed. Not running target
Error: Process completed with exit code 1.
```

Judging from the error message and [the diff](https://github.com/uber/hermetic_cc_toolchain/compare/v3.1.1...v3.2.0), I guess the non-empty `root_module_direct_deps` introduced in `3.2.0` [here](https://github.com/uber/hermetic_cc_toolchain/blob/bfc407599e503a44928a3cefad27421c9341eff0/toolchain/ext.bzl#L23) started to disallow the users to use the module extension as a dev dependency.

```
# In my MODULE.bazel using `hermetic_cc_toolchain`
cc_toolchains = use_extension(
    "@hermetic_cc_toolchain//toolchain:ext.bzl",
    "toolchains",
    dev_dependency = True, # If this line is removed, the error disappears
)
```

I'd like the toolchain to be available even as a dev dependency.

--- a/MODULE.bazel
+++ b/MODULE.bazel
@@ -30,7 +30,7 @@ use_repo(
     "com_github_tetratelabs_wazero",
 )
 
-toolchains = use_extension("//toolchain:ext.bzl", "toolchains")
+toolchains = use_extension("//toolchain:ext.bzl", "toolchains", dev_dependency = True)
 use_repo(toolchains, "zig_sdk")
 
 # TODO: register all toolchains by:

--- a/toolchain/defs.bzl
+++ b/toolchain/defs.bzl
@@ -95,7 +95,7 @@ def toolchains(
         host_platform_ext = host_platform_ext,
     )
 
-    indirect_repos = ["zig_config"]
+    private_repos = ["zig_config"]
 
     for os, archs in exec_platforms.items():
         for arch in archs:
@@ -108,11 +108,11 @@ def toolchains(
                 exec_os = os,
                 exec_arch = arch,
             )
-            indirect_repos.append("zig_config-{}-{}".format(os, arch))
+            private_repos.append("zig_config-{}-{}".format(os, arch))
 
     return struct(
-        direct = ["zig_sdk"],
-        indirect = indirect_repos,
+        public = ["zig_sdk"],
+        private = private_repos,
     )
 
 def _quote(s):

--- a/toolchain/ext.bzl
+++ b/toolchain/ext.bzl
@@ -17,23 +17,30 @@ _exec_platform = tag_class(
 
 def _toolchains_impl(mctx):
     exec_platforms = {}
+    root_direct_deps = []
+    root_direct_dev_deps = []
+    is_non_dev_dependency = mctx.root_module_has_non_dev_dependency
 
     for mod in mctx.modules:
-        for ep in mod.tags.exec_platform:
-            _archs = exec_platforms.get(ep.os, list())
-            if ep.arch not in _archs:
-                _archs.append(ep.arch)
-            exec_platforms[ep.os] = _archs
+        if mod.is_root:
+            for ep in mod.tags.exec_platform:
+                _archs = exec_platforms.get(ep.os, list())
+                if ep.arch not in _archs:
+                    _archs.append(ep.arch)
+                exec_platforms[ep.os] = _archs
 
-    repos = zig_toolchains(exec_platforms = exec_platforms)
+            repos = zig_toolchains(exec_platforms = exec_platforms)
+
+            root_direct_deps = list(repos.public) if is_non_dev_dependency else []
+            root_direct_dev_deps = list(repos.public) if not is_non_dev_dependency else []
 
     metadata_kwargs = {}
     if bazel_features.external_deps.extension_metadata_has_reproducible:
         metadata_kwargs["reproducible"] = True
 
     return mctx.extension_metadata(
-        root_module_direct_deps = repos.direct,
-        root_module_direct_dev_deps = [],
+        root_module_direct_deps = root_direct_deps,
+        root_module_direct_dev_deps = root_direct_dev_deps,
         **metadata_kwargs
     )
 

